src/server/ua_server_binary.c

```diff
@@ -647,15 +647,18 @@ createSecureChannel(void *application, UA_Connection *connection,
     UA_Server *server = (UA_Server*)application;
 
     /* Iterate over available endpoints and choose the correct one */
-    UA_Endpoint *endpoint = NULL;
+    UA_EndpointDescription *endpoint = NULL;
+    UA_SecurityPolicy *securityPolicy = NULL;
     UA_StatusCode retval = UA_STATUSCODE_GOOD;
     for(size_t i = 0; i < server->config.endpointsSize; ++i) {
-        UA_Endpoint *endpointCandidate = &server->config.endpoints[i];
+        UA_EndpointDescription *endpointCandidate = &server->config.endpoints[i];
         if(!UA_ByteString_equal(&asymHeader->securityPolicyUri,
-                                &endpointCandidate->securityPolicy.policyUri))
+                                &endpointCandidate->securityPolicyUri))
             continue;
-        retval = endpointCandidate->securityPolicy.asymmetricModule.
-            compareCertificateThumbprint(&endpointCandidate->securityPolicy,
+        securityPolicy = UA_SecurityPolicy_getSecurityPolicyByUri(server,
```

Comment:
Maybe it would be better to make the certificate comparison a part of the get function?
Because a specific policy instance is uniquely identified by the endpoint and the certificate.

Also, maybe it would be best to make a policyManager plugin(?) so that we can add or remove policys on the fly?

Reply:
> Maybe it would be better to make the certificate comparison a part of the get function?
> Because a specific policy instance is uniquely identified by the endpoint and the certificate.

I tried to find any statement in the specification that confirms this but without success. Do you mean that a server can have several instances of a SecurityPolicy with different certificates? The only thing I can find in the specification is in Part 4, Table 11 - OpenSecureChannel Service Parameters: "The SecurityPolicy is identified by the securityPolicyUri". The idea with this PR is to be able to use the SecurityPolicies also in the access control plugin to decrypt passwords. In the ActivateSessionRequest no ReceiverCertificateThumbprint is included.

> Also, maybe it would be best to make a policyManager plugin(?) so that we can add or remove policys on the fly?

Sure, but what is the use case to add and remove the SecurityPolicies on the fly? Can you share some more details of how this manager should look like?

Reply:
There are cases where we can have several "competing" policies with the same URI.

- If the server is visible in different networks (towards the internet and internally) with different certificates.
- When the GDS updates the server certificate, there could be some time with parallel policy instances.

Some if this stuff is far-fetched. Lets keep this PR simple.
And add this to the lookup when the need arises.